rclpy/rclpy/action/client.py

```diff
@@ -209,7 +209,10 @@ def _remove_pending_request(self, future, pending_requests):
     def _remove_pending_goal_request(self, future):
         seq = self._remove_pending_request(future, self._pending_goal_requests)
         if seq in self._sequence_number_to_goal_id:
+            goal_uuid = bytes(self._sequence_number_to_goal_id[seq].uuid)
             del self._sequence_number_to_goal_id[seq]
+            if goal_uuid in self._feedback_callbacks:
+                del self._feedback_callbacks[goal_uuid]
```

Comment:
this changes behavior a bit. previously feedback callback still can be issued after corresponding goal has been completed. but with this fix, that cannot be done. I am not sure if this behavior is intention or not.

Reply:
what's the behavior in rclcpp?

Reply:
thanks for asking, `rclcpp` could be reference.

if user is NOT aware of result (either result callback option is not provided or async_get_result is not called), feedback callback can be issued as long as user application keeps reference to `std::shared_future<typename GoalHandle::SharedPtr>`.

https://github.com/ros2/rclcpp/blob/c24e485084f421292734ebf6b63f1a105a2c7602/rclcpp_action/include/rclcpp_action/client.hpp#L421-L422

https://github.com/ros2/rclcpp/blob/c24e485084f421292734ebf6b63f1a105a2c7602/rclcpp_action/include/rclcpp_action/client.hpp#L651-L662

if user is aware of result in anyways, then once result is received from action server to client, `goal_handles_[goal_id]` is erased. in this case, feedback callback will not be called anymore.

https://github.com/ros2/rclcpp/blob/c24e485084f421292734ebf6b63f1a105a2c7602/rclcpp_action/include/rclcpp_action/client.hpp#L727

current code is totally wrong, it should be removed aligned with `_pending_result_requests`.

Reply:
I will fix the code soon.

Reply:
> if user is NOT aware of result (either result callback option is not provided or async_get_result is not called), feedback callback can be issued as long as user application keeps reference to std::shared_future<typename GoalHandle::SharedPtr>.

Maybe, we could keep a weak reference to `ClientGoalHandle`  in https://github.com/ros2/rclpy/blob/f1f40d6cd42bcf1988b2c72c8d0ce03bc62ec530/rclpy/rclpy/action/client.py#L167
to get something equivalent (?).
I'm not sure.

Though removing the feedback callback when the result is get is also not a bad idea.

@jacobperron @sloretz @adityapande-1995 any opinions?

Reply:
Removing the feedback callback after the goal is completed seems reasonable to me. I don't think there's anything in the design document or other documentation that indicates we should expect feedback after a goal is completed.

Reply:
However, are we interpreting this future correctly? I thought that receiving the response for the goal request means that the goal was accepted, not necessarily completed. We don't know that the goal is completed unless we check the status or the result is received.